src/util/amount.rs

```diff
@@ -0,0 +1,1336 @@
+    } else if denom.precision() < 0 {
+        // need to inject a comma in the number
+        let nb_decimals = denom.precision().abs() as usize;
+        let real = format!("{:0width$}", satoshi, width = nb_decimals);
+        if real.len() == nb_decimals {
+            write!(f, "0.{}", &real[real.len() - nb_decimals..])?;
+        } else {
+            write!(
+                f,
+                "{}.{}",
+                &real[0..(real.len() - nb_decimals)],
+                &real[real.len() - nb_decimals..]
+            )?;
+        }
+    } else {
+        // denom.precision() == 0
+        write!(f, "{}", satoshi)?;
+    }
+    Ok(())
+}
+
+/// Amount
+///
+/// The [Amount] type can be used to express Bitcoin amounts that supports
+/// arithmetic and convertion to various denominations.
+///
+///
+/// Warning!
+///
+/// This type implements several arithmetic operations from [std::ops].
+/// To prevent errors due to overflow or underflow when using these operations,
+/// it is advised to instead use the checked arithmetic methods whose names
+/// start with `checked_`.  The operations from [std::ops] that [Amount]
+/// implements will panic when overflow or underflow occurs.  Also note that
+/// since the internal representation of amounts is unsigned, subtracting below
+/// zero is considered an underflow and will cause a panic if you're not using
+/// the checked arithmetic methods.
+///
+#[derive(Copy, Clone, Hash)]
+pub struct Amount(u64);
+
+impl Amount {
+    /// The zero amount.
+    pub const ZERO: Amount = Amount(0);
+    /// Exactly one satoshi.
+    pub const ONE_SAT: Amount = Amount(1);
+    /// Exactly one bitcoin.
+    pub const ONE_BTC: Amount = Amount(100_000_000);
+
+    /// Create an [Amount] with satoshi precision and the given number of satoshis.
+    pub fn from_sat(satoshi: u64) -> Amount {
+        Amount(satoshi)
+    }
+
+    /// Get the number of satoshis in this [Amount].
+    pub fn as_sat(self) -> u64 {
+        self.0
+    }
+
+    /// The maximum value of an [Amount].
+    pub fn max_value() -> Amount {
+        Amount(u64::max_value())
+    }
+
+    /// The minimum value of an [Amount].
+    pub fn min_value() -> Amount {
+        Amount(u64::min_value())
+    }
+
+    /// Convert from a value expressing bitcoins to an [Amount].
+    pub fn from_btc(btc: f64) -> Result<Amount, ParseAmountError> {
+        Amount::from_float_in(btc, Denomination::Bitcoin)
+    }
+
+    /// Parse a decimal string as a value in the given denomination.
+    ///
+    /// Note: This only parses the value string.  If you want to parse a value
+    /// with denomination, use [FromStr].
+    pub fn from_str_in(s: &str, denom: Denomination) -> Result<Amount, ParseAmountError> {
+        let (negative, satoshi) = parse_signed_to_satoshi(s, denom)?;
+        if negative {
+            return Err(ParseAmountError::Negative);
+        }
+        if satoshi > i64::max_value() as u64 {
```

Comment:
Why do we need to limit the range here (not that it had any practical implications, just wondering)?

Reply:
Presumably to keep the range of `Amount` and `SignedAmount` the same?

Reply:
IIRC I didn't want to have a flag saying "this is intended for SignedAmount". If it's larger than that, it would fit in an Amount, but for SignedAmount if you did `returnvalue as i64` it'd overflow and be negative. Since 21 million coins < 2**63 anyway, it shouldn't matter.

Reply:
Oh I just realize this check is not in the helper method but in the `Amount` method. Yeah doesn't have to be there at all. I think it's there because of copy pasting. You can PR to remove it if it bothers you.